### ISSUE TITLE
Reentrant locks in CDI contexts can leak Thread instances into image heap

### ISSUE THREAD

**Issue by galderz** [quarkusio/quarkus#40737]
### Describe the bug

Native builds can fail with errors like this:

```
Error: Detected a started Thread in the image heap. Thread name: main. Threads running in the image generator are no longer running at image runtime. If these objects should not be stored in the image heap, you can use

    '--trace-object-instantiation=java.lang.Thread'

to find classes that instantiate these objects. Once you found such a class, you can mark it explicitly for run time initialization with

    '--initialize-at-run-time=<culprit>'

to prevent the instantiation of the object.
The object was probably created by a class initializer and is reachable from a static field. You can request class initialization at image runtime by using the option --initialize-at-run-time=<class-name>. Or you can write your own initialization methods and call them explicitly from your main entry point.
Detailed message:
Trace: Object was reached by
  reading field java.util.concurrent.locks.AbstractOwnableSynchronizer.exclusiveOwnerThread of constant
    java.util.concurrent.locks.ReentrantLock$NonfairSync@69557381: java.util.concurrent.locks.ReentrantLock$NonfairSync@69557381[State = 1, empty q...
  reading field java.util.concurrent.locks.ReentrantLock.sync of constant
    java.util.concurrent.locks.ReentrantLock@4c3d230a: java.util.concurrent.locks.ReentrantLock@4c3d230a[Locked by thread main]
  reading field io.quarkus.arc.impl.LazyValue.lock of constant
    io.quarkus.arc.impl.LazyValue@57f1e065: io.quarkus.arc.impl.LazyValue@57f1e065
  reading field java.util.concurrent.ConcurrentHashMap$Node.val of constant
    java.util.concurrent.ConcurrentHashMap$Node@77437480: 9SDuiV8v87J2ZUi6goJuuKNan1k=io.quarkus.arc.impl.LazyValue@57f1e065
  indexing into array java.util.concurrent.ConcurrentHashMap$Node[]@e0ece13: [Ljava.util.concurrent.ConcurrentHashMap$Node;@e0ece13
  reading field java.util.concurrent.ConcurrentHashMap.table of constant
    java.util.concurrent.ConcurrentHashMap@6ea31905: {fRqO007voVblR_BoQkAFpaGPHtI=io.quarkus.arc.impl.LazyValue@7714042a, 9SDuiV8v87J...
  reading field io.quarkus.arc.impl.ComputingCache.map of constant
    io.quarkus.arc.impl.ComputingCache@1b558c4d: io.quarkus.arc.impl.ComputingCache@1b558c4d
  reading field io.quarkus.arc.impl.ComputingCacheContextInstances.instances of constant
    io.quarkus.arc.impl.ComputingCacheContextInstances@18dd2ee4: io.quarkus.arc.impl.ComputingCacheContextInstances@18dd2ee4
  reading field io.quarkus.arc.impl.AbstractSharedContext.instances of constant
    io.quarkus.arc.impl.SingletonContext@68bc232f: io.quarkus.arc.impl.SingletonContext@68bc232f
  reading field java.util.Collections$SingletonList.element of constant
    java.util.Collections$SingletonList@145b22a5: [io.quarkus.arc.impl.SingletonContext@68bc232f]
  reading field io.quarkus.arc.impl.Contexts.singletonContextSingleton of constant
    io.quarkus.arc.impl.Contexts@4cea0bac: io.quarkus.arc.impl.Contexts@4cea0bac
  reading field io.quarkus.arc.impl.ArcContainerImpl.contexts of constant
    io.quarkus.arc.impl.ArcContainerImpl@5283e01d: ArcContainerImpl [id=1, running=true, beans=50, observers=14, scopes=4]
  reading field java.util.concurrent.atomic.AtomicReference.value of constant
    java.util.concurrent.atomic.AtomicReference@2bdff16d: ArcContainerImpl [id=1, running=true, beans=50, observers=14, scopes=4]
  scanning root java.util.concurrent.atomic.AtomicReference@2bdff16d: ArcContainerImpl [id=1, running=true, beans=50, observers=14, scopes=4] embedded in
    io.quarkus.arc.Arc.container(Arc.java:56)
  parsing method io.quarkus.arc.Arc.container(Arc.java:56) reachable via the parsing context
    at io.quarkus.arc.impl.ArcCDIProvider.getCDI(ArcCDIProvider.java:28)
    at jakarta.enterprise.inject.spi.CDI.checkProvider(CDI.java:103)
    at jakarta.enterprise.inject.spi.CDI$$Lambda/0x00000007c1724f20.test(Unknown Source)
    at java.util.concurrent.CopyOnWriteArrayList.bulkRemove(CopyOnWriteArrayList.java:925)
    at java.util.concurrent.CopyOnWriteArrayList$COWSubList.bulkRemove(CopyOnWriteArrayList.java:1642)
    at java.util.concurrent.CopyOnWriteArrayList$COWSubList.removeAll(CopyOnWriteArrayList.java:1626)
    at root method.(Unknown Source)
```

The thread here is the `main` thread, which is a thread expected to be part of the native image execution, but the real issue here is that image data contains references to this thread. These references are stored inside `ReentrantLock` instances (`exclusiveOwnerThread` field) inside `io.quarkus.arc.impl.LazyValue`. These are in turn stored within a `ConcurrentHashMap` inside `ComputingCache` as part of `io.quarkus.arc.impl.SingletonContext` for the CDI contexts.

Assuming the concurrency requirements are needed due to the multi-threaded nature of native image builds, Quarkus needs to make sure the concurrency aspects go away. IOW, the cache in `SingletonContext` should probably become an immutable hashmap of values that have no lock references. An easier fix here might be to make sure that all the locks inside the `LazyValue` instances are released. The tricky aspect is when to do so? As image heap objects are created, checks are made to make sure no forbidden types end up in the image heap. So ideally these values should become immutable as soon as they've been computed and only store immutable values.

Other solutions might exist, such as: is the concurrent access/update required? Can these `ContextInstance` objects need to be in the image heap?

**Comment by mkouba** [comment]
Hm, do we know why it only happens occasionally? 

> An easier fix here might be to make sure that all the locks inside the LazyValue instances are released.

All the locks should be released. If they're not then we should try to find out why...

We use the same strategy (CHM + `LazyValue`) for the application context BTW.

FTR previously, the `LazyValue` used `volatile` + `synchronized` but we switched to Lock to become virtual threads-friendly.

**Comment by galderz** [comment]
If you need locking, I would recommend trying to use a lock that does not keep a strong reference to a Thread. Could something be constructed around thread ids instead? /cc @franz1981 

The exception above shows that the lock has not been released:

```
    java.util.concurrent.locks.ReentrantLock$NonfairSync@69557381: java.util.concurrent.locks.ReentrantLock$NonfairSync@69557381[State = 1, empty q...
```

The `State = 1...` part means that a lock has been acquired but it has not yet been released. If it had been released, state would be 0 as per the ReentrantLock/AbstractQueuedSynchronizer implementations.

How is it possible that this happens? Each points-to-analysis iteration discovers objects that are referenced from static objects, e.g. from `Arc.INSTANCE`, it scans its fields to find new objects..., and during this process an image objects will be created for all these objects. One of the things done when creating image objects is checking for disallowed image heap objects, e.g. `Thread`.

The chances of the bug here happening are small but it can happen. For the issue to appear, points-to-analysis needs to be try to create image object for the lock inside the `LazyValue`. The window is quite small since it needs to happen within the window that the lock is held in `LazyValue.get()`:

```
public T get() {
    // ...

    lock.lock();
    // Between here...
    try {
        if (value == null) {
            value = supplier.get();
        }
        return value;
    } finally {
        lock.unlock();
        // And here
    }
}
```

I can't think of anything that really can be done at the Quarkus level to protect against this /cc @zakkak

**Comment by mkouba** [comment]
> I can't think of anything that really can be done at the Quarkus level to protect against this kind of race condition. Avoiding strong references to disallowed instances is the best option I can think of /cc @zakkak

Are you saying that any extension that makes use of `ReentrantLock` during the `STATIC_INIT` can break native image compilation? :eyes:

**Comment by dmlloyd** [comment]
Locking during static init should be fine, however the lock should be definitely released using `finally` and should not have some chance to be re-acquired during the "steady state" at the end of native image build. I would think that if this invariant is respected, you should have no trouble.

**Comment by galderz** [comment]
@mkouba @dmlloyd I don't think the issue is necessarily about the reentrant lock acquire/release, but rather the fact that the reentrant lock is part of the image heap state (via `Arc.INSTANCE`), and we can't control when `Arc.INSTANCE` and any references from that will become image heap objects, at which stage the verifications for invalid objects is carried out.

**Comment by zakkak** [comment]
> and we can't control when `Arc.INSTANCE` and any references from that will become image heap objects, at which stage the verifications for invalid objects is carried out.

That doesn't sound right to me. Isn't the image heap supposed to get checked/verified after all iterations have been done? I.e. when the heap is in a stable state and there is no build-time code being run anymore?

**Comment by dmlloyd** [comment]
@galderz is there any way you can get a look at the fields of the exact object that has the strong reference to the thread?

**Comment by dmlloyd** [comment]
If it is in fact `java.util.concurrent.locks.AbstractOwnableSynchronizer.exclusiveOwnerThread` then the lock was simply not released for some reason. That field is nulled out on release AFAICT.

**Comment by mkouba** [comment]
> IOW, the cache in SingletonContext should probably become an immutable hashmap of values that have no lock references.

@galderz This cannot be done. The singleton context is not prepopulated during the STATIC_INIT. Beans are instantiated lazily when first used, i.e. at any time after the application starts.

> An easier fix here might be to make sure that all the locks inside the LazyValue instances are released.

"released" as in calling `Lock#unlock()`, or?

>>    and we can't control when Arc.INSTANCE and any references from that will become image heap objects, at which stage the verifications for invalid objects is carried out.

> That doesn't sound right to me. Isn't the image heap supposed to get checked/verified after all iterations have been done? I.e. when the heap is in a stable state and there is no build-time code being run anymore?

@zakkak I would expect this behavior as well.

> If it is in fact `java.util.concurrent.locks.AbstractOwnableSynchronizer.exclusiveOwnerThread` then the lock was simply not released for some reason. That field is nulled out on release AFAICT.

@dmlloyd Yes, it seems to be the `java.util.concurrent.locks.AbstractOwnableSynchronizer.exclusiveOwnerThread` (at least that's what I read in the original description). And yes, it shoud be nulled out in `ReentrantLock.Sync.tryRelease(int)`.

**Comment by galderz** [comment]
> > and we can't control when `Arc.INSTANCE` and any references from that will become image heap objects, at which stage the verifications for invalid objects is carried out.
>
> That doesn't sound right to me. Isn't the image heap supposed to get checked/verified after all iterations have been done? I.e. when the heap is in a stable state and there is no build-time code being run anymore?

That's what I thought initially to happen but having looked at the code it seems to be object reachable callbacks, on which thread detection is based, can happen at any time. Have you seen something in the code that makes you think that? Let's stop going around in circles about this.

Furthermore, here's output on the main thread being checked to see if it's reachable. There's no indication that this is happening at the end of the points-to analysis and indeed it pops even before the first points-to iteration has finished (which is signalled showing the '*' char). To reiterate, I don't see any guarantees on when this check will happen, so it can indeed, through timing such that main thread is marked reacheable while another thread is in the middle of the lock acquire/release block that I mentioned above. This is consistent with the behaviour of the bug, which happens randomly and therefore the right timing needs to happen.

```
Build resources:
 - 24.18GB of memory (75.6% of 32.00GB system memory, determined at start)
 - 10 thread(s) (100.0% of 10 available processor(s), determined at start)
Check for thread to be reachable: Thread[#38,main,5,main]
Stacktrace:
java.lang.Exception
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.image.DisallowedImageHeapObjects.onThreadReachable(DisallowedImageHeapObjects.java:122)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.image.DisallowedImageHeapObjectFeature.lambda$duringSetup$4(DisallowedImageHeapObjectFeature.java:79)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.meta.AnalysisType.lambda$notifyObjectReachable$12(AnalysisType.java:659)
	at java.base/java.util.concurrent.ConcurrentHashMap$KeySetView.forEach(ConcurrentHashMap.java:4714)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.util.ConcurrentLightHashSet.forEach(ConcurrentLightHashSet.java:146)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.meta.AnalysisType.notifyObjectReachable(AnalysisType.java:659)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.heap.ImageHeapScanner.onObjectReachable(ImageHeapScanner.java:586)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.heap.SVMImageHeapScanner.onObjectReachable(SVMImageHeapScanner.java:125)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.heap.ImageHeapScanner.lambda$markReachable$10(ImageHeapScanner.java:572)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.util.CompletionExecutor.executeCommand(CompletionExecutor.java:166)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.util.CompletionExecutor.lambda$executeService$0(CompletionExecutor.java:152)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.compute(ForkJoinTask.java:1726)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.compute(ForkJoinTask.java:1717)
	at java.base/java.util.concurrent.ForkJoinTask$InterruptibleTask.exec(ForkJoinTask.java:1641)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:507)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1489)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:2071)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:2033)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:189)
12:15:03,625 WARN  [io.net.res.dns.DnsServerAddressStreamProviders] Can not find io.netty.resolver.dns.macos.MacOSDnsServerAddressStreamProvider in the classpath, fallback to system defaults. This may result in incorrect DNS resolutions on MacOS. Check whether you have a dependency on 'io.netty:netty-resolver-dns-native-macos'
[2/8] Performing analysis...  [*****]
```

I have tried to introduced a small delay inside the LazyValye.get() acquire/release section but have not yet reproduced the issue. I'll try to increase it and see if it breaks.

**Comment by zakkak** [comment]
> That's what I thought initially would happen, but having looked at the code it used object reachable callbacks to detect thread leaks and those callbacks can happen at any time. Have you seen something in the code that makes you think that?

I had a look and did some testing. So what GraalVM ensures [is that sequential class initialization will complete before proceeding with the analysis](https://github.com/oracle/graal/blob/00e298ceeeabaae1375b7889ba67cbae4a1ec7b6/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/classinitialization/ClassInitializationSupport.java#L171-L176). If the class initialization ends up deadlocking or taking way too long GraalVM's wathdog should detect it and fail the build, which doesn't appear to be the case here.

Now if we add concurrency in class initialization there is no way for GraalVM to detect whether all threads have completed or not. That said if a class initializer spawns a thread and returns, the spawned thread can potentially run in parallel with the analysis (e.g. the reachable callbacks) and image building in general which can be problematic...

What's interesting in this specific case though is that the thread holding the lock appears to be `main`, i.e. the sequential part of the class initializer has still not released the lock. 

Another case where class initialization happens during Analysis, and might thus result in what we see above, is when a type not already initialized is [found to be reachable by the analysis](https://github.com/oracle/graal/blob/e69b8a71be11b1ec0f7fc7aed969ae82db48bf54/substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/meta/AnalysisUniverse.java#L699). In this case class initialization is triggered by whichever thread happened to detect the reachable type, and the initialization runs in parallel with other callbacks which could lead to the observed behavior...

So to sum up I believe that we need to detect which class initializer is triggering the LazyValue instance "initialization" (i.e. acquires the lock) and ensure that it is either no longer run or it runs before the analysis phase.

I will try to have another look the next days.

**Comment by galderz** [comment]
Thanks for the investigation @zakkak 

> So to sum up I believe that we need to detect which class initializer is triggering the LazyValue instance "initialization" (i.e. acquires the lock) and ensure that it is either no longer run or it runs before the analysis phase.

I'm not sure how feasible this is, but looking at how objects are added to `ComputingCacheContextInstances`, these contexts are added as a result of creating CDI beans. In the original exception:

```
    java.util.concurrent.ConcurrentHashMap$Node@77437480: 9SDuiV8v87J2ZUi6goJuuKNan1k=io.quarkus.arc.impl.LazyValue@57f1e065
```

`9SDuiV8v87J2ZUi6goJuuKNan1k` is the bean ID. If the way these beans are added to context can be somehow controlled to avoid the situation you explain, then maybe it can work. But if these beans are added to context as user types are discovered, then it might be hard to do. Since beans are discovered through annotations and we have the jandex for it, in theory all bean discovery should be doable early on and in a single pass.

Using something other than ReentrantLock in `LazyValue` might be easier to implement/achieve, but the `Thread` issue here could manifest in some other way with say, another forbidden type, so I can see the value of exploring this further.

**Comment by zakkak** [comment]
So far I see that:

1. `io.quarkus.runner.ApplicationImpl` [is registered for reflection](https://github.com/quarkusio/quarkus/blob/4c2550a8377905d6c3c9086a3970cdbbe9a58384/core/deployment/src/main/java/io/quarkus/deployment/steps/MainClassBuildStep.java#L513-L516)
2. This triggers an `onTypeReachable` callback which triggers the class initialization which ends up acquiring the lock


<details>
<summary>The stack trace at this point looks like this:</summary>

```java
get:31, LazyValue (io.quarkus.arc.impl)
beanManager:404, ArcContainerImpl (io.quarkus.arc.impl)
<init>:40, ArcCDIProvider$ArcCDI (io.quarkus.arc.impl)
<init>:23, ArcCDIProvider (io.quarkus.arc.impl)
init:273, ArcContainerImpl (io.quarkus.arc.impl)
initialize:40, Arc (io.quarkus.arc)
initContainer:49, ArcRecorder (io.quarkus.arc.runtime)
deploy_0:-1, ArcProcessor$initializeContainer1770303700 (io.quarkus.deployment.steps)
deploy:-1, ArcProcessor$initializeContainer1770303700 (io.quarkus.deployment.steps)
<clinit>:-1, ApplicationImpl (io.quarkus.runner)
ensureClassInitialized0:-1, Unsafe (jdk.internal.misc)
ensureClassInitialized:1160, Unsafe (jdk.internal.misc)
ensureClassInitialized:177, ClassInitializationSupport (com.oracle.svm.hosted.classinitialization)
computeInitKindAndMaybeInitializeClass:191, AllowAllHostedUsagesClassInitializationSupport (com.oracle.svm.hosted.classinitialization)
computeInitKindAndMaybeInitializeClass:129, AllowAllHostedUsagesClassInitializationSupport (com.oracle.svm.hosted.classinitialization)
maybeInitializeAtBuildTime:161, ClassInitializationSupport (com.oracle.svm.hosted.classinitialization)
maybeInitializeAtBuildTime:150, ClassInitializationSupport (com.oracle.svm.hosted.classinitialization)
onTypeReachable:310, SVMHost (com.oracle.svm.hosted)
onTypeReachable:699, AnalysisUniverse (com.oracle.graal.pointsto.meta)
lambda$new$0:310, AnalysisType (com.oracle.graal.pointsto.meta)
run:-1, AnalysisType$$Lambda/0x0000000801102dd0 (com.oracle.graal.pointsto.meta)
call:572, Executors$RunnableAdapter (java.util.concurrent)
run:317, FutureTask (java.util.concurrent)
ensureDone:63, AnalysisFuture (com.oracle.graal.pointsto.util)
ensureOnTypeReachableTaskDone:696, AnalysisType (com.oracle.graal.pointsto.meta)
onReachable:590, AnalysisType (com.oracle.graal.pointsto.meta)
run:-1, AnalysisType$$Lambda/0x0000000801103660 (com.oracle.graal.pointsto.meta)
atomicSetAndRun:49, AtomicUtils (com.oracle.graal.pointsto.util)
lambda$registerAsReachable$8:562, AnalysisType (com.oracle.graal.pointsto.meta)
accept:-1, AnalysisType$$Lambda/0x0000000801103430 (com.oracle.graal.pointsto.meta)
forAllSuperTypes:676, AnalysisType (com.oracle.graal.pointsto.meta)
forAllSuperTypes:659, AnalysisType (com.oracle.graal.pointsto.meta)
forAllSuperTypes:655, AnalysisType (com.oracle.graal.pointsto.meta)
registerAsReachable:562, AnalysisType (com.oracle.graal.pointsto.meta)
registerClass:202, ReflectionDataBuilder (com.oracle.svm.hosted.reflect)
lambda$register$1:171, ReflectionDataBuilder (com.oracle.svm.hosted.reflect)
run:-1, ReflectionDataBuilder$$Lambda/0x0000000801133b70 (com.oracle.svm.hosted.reflect)
executeCommand:187, CompletionExecutor (com.oracle.graal.pointsto.util)
lambda$executeService$0:171, CompletionExecutor (com.oracle.graal.pointsto.util)
run:-1, CompletionExecutor$$Lambda/0x00000008011b1168 (com.oracle.graal.pointsto.util)
exec:1423, ForkJoinTask$RunnableExecuteAction (java.util.concurrent)
doExec:387, ForkJoinTask (java.util.concurrent)
externalHelpQuiesce:2260, ForkJoinPool (java.util.concurrent)
helpQuiescePool:2297, ForkJoinPool (java.util.concurrent)
awaitQuiescence:3565, ForkJoinPool (java.util.concurrent)
complete:237, CompletionExecutor (com.oracle.graal.pointsto.util)
doTypeflow:529, PointsToAnalysis (com.oracle.graal.pointsto)
finish:517, PointsToAnalysis (com.oracle.graal.pointsto)
runAnalysis:162, AbstractAnalysisEngine (com.oracle.graal.pointsto)
runPointsToAnalysis:783, NativeImageGenerator (com.oracle.svm.hosted)
doRun:592, NativeImageGenerator (com.oracle.svm.hosted)
run:550, NativeImageGenerator (com.oracle.svm.hosted)
buildImage:538, NativeImageGeneratorRunner (com.oracle.svm.hosted)
build:720, NativeImageGeneratorRunner (com.oracle.svm.hosted)
start:142, NativeImageGeneratorRunner (com.oracle.svm.hosted)
main:97, NativeImageGeneratorRunner (com.oracle.svm.hosted)
```

</details>

<details>
<summary>
Note that I even managed to get the breakpoint to hit while "main" was indeed holding the lock:
</summary>

```
lock = {ReentrantLock@21401} "java.util.concurrent.locks.ReentrantLock@575d8c53[Locked by thread main]"
 sync = {ReentrantLock$NonfairSync@22299} "java.util.concurrent.locks.ReentrantLock$NonfairSync@31c5a544[State = 1, empty queue]"
  head = null
  tail = null
  state = 1
  exclusiveOwnerThread = {Thread@1} "Thread[#1,main,5,main]"
   eetop = 140043113163408
   tid = 1
   name = "main"
   interrupted = false
   contextClassLoader = {NativeImageClassLoader@22303} 
   inheritedAccessControlContext = {AccessControlContext@22304} 
   holder = {Thread$FieldHolder@22305} 
   threadLocals = {ThreadLocal$ThreadLocalMap@22306} 
   inheritableThreadLocals = null
   scopedValueBindings = {Class@885} "class java.lang.Thread"
   interruptLock = {Object@22307} 
   parkBlocker = null
   nioBlocker = null
   cont = null
   uncaughtExceptionHandler = null
   threadLocalRandomSeed = 6170759035374141833
   threadLocalRandomProbe = -1463423431
   threadLocalRandomSecondarySeed = 1232312558
   container = null
   headStackableScopes = null
```

</details>

Now at this point, if `io.quarkus.arc.Arc.container` happens to get parsed (by another thread) while `main` still has the lock it could lead to the observed error.

@galder what Mandrel/GraalVM version were you able to reproduce this with? If you can only reproduce with `master` builds, it could be https://github.com/oracle/graal/commit/9d52cb92bf441c2c9ab0a2c628667dcd52985986 causing this.

**Comment by galderz** [comment]
@zakkak I didn't reproduce this myself, @gsmet found it in the failures in https://github.com/quarkiverse/quarkus-github-app/actions/runs/9153770878/job/25163260450 It was using Mandrel-23.1.3.1-Final

**Comment by zakkak** [comment]
OK, but the trace in https://github.com/quarkusio/quarkus/issues/40737#issuecomment-2152252430 is not from Mandrel-23.1.3.1-Final, right? How did you get that trace? Did you use a synthetic "reproducer"?

**Comment by zakkak** [comment]
> Another case where class initialization happens during Analysis, and might thus result in what we see above, is when a type not already initialized is [found to be reachable by the analysis](https://github.com/oracle/graal/blob/e69b8a71be11b1ec0f7fc7aed969ae82db48bf54/substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/meta/AnalysisUniverse.java#L699). In this case class initialization is triggered by whichever thread happened to detect the reachable type, and the initialization runs in parallel with other callbacks which could lead to the observed behavior...

To add some more info to this... 

Although a class initialization might get triggered this way [GraalVM tries to ensure that it won't have any side effects](https://github.com/graalvm/mandrel/blob/mandrel-23.1.3.1-Final/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/classinitialization/EarlyClassInitializerAnalysis.java#L170-L209) before doing so. According to these checks if a class initializer accesses any static field from a different class the class initialization is postponed for run-time, so accessing the reentrant lock should not be possible this way, unless there is some bug in the checks and the code escapes them (but we can't tell with the current data).

**Comment by galderz** [comment]
> OK, but the trace in [#40737 (comment)](https://github.com/quarkusio/quarkus/issues/40737#issuecomment-2152252430) is not from Mandrel-23.1.3.1-Final, right? How did you get that trace? Did you use a synthetic "reproducer"?

Ah yes, this particular comment was produced on master, not 23.1.3.1.

**Comment by mkouba** [comment]
@galderz @zakkak What's the status of this issue guys?

**Comment by zakkak** [comment]
Unfortunately this has been on freeze for quite some time now. I will need to get back to it. Thanks for the reminder.

**Comment by zakkak** [comment]
I am going to close this as `won't fix`. I am not able to reproduce the issue and it looks like it never happened again in the https://github.com/quarkiverse/quarkus-github-app/actions/workflows/quarkus-snapshot.yaml workflow where it first appeared.

FTR I am appending my so far analysis that concludes in that Quarkus doesn't seem to do anything wrong that could lead to that behavior.

The issue indicates that `io.quarkus.arc.Arc.container` parsing (which was reached through `io.quarkus.arc.impl.ArcCDIProvider#getCDI`) fails due to the reentrant lock of `io.quarkus.arc.impl.BeanManagerImpl#INSTANCE` already being held by another thread (`main` in our case)

So how/when is `io.quarkus.arc.impl.BeanManagerImpl#INSTANCE` instantiated and how/when is `io.quarkus.arc.Arc.container` first parsed.

<details>
<summary>`io.quarkus.arc.impl.BeanManagerImpl#INSTANCE` in my tests is first instantiated through:</summary>

```
<clinit>:54, BeanManagerImpl (io.quarkus.arc.impl)
beanManager:404, ArcContainerImpl (io.quarkus.arc.impl)
<init>:40, ArcCDIProvider$ArcCDI (io.quarkus.arc.impl) // <----------------------
<init>:23, ArcCDIProvider (io.quarkus.arc.impl)
init:273, ArcContainerImpl (io.quarkus.arc.impl)
initialize:40, Arc (io.quarkus.arc)
initContainer:49, ArcRecorder (io.quarkus.arc.runtime)
deploy_0:-1, ArcProcessor$initializeContainer1770303700 (io.quarkus.deployment.steps)
deploy:-1, ArcProcessor$initializeContainer1770303700 (io.quarkus.deployment.steps)
<clinit>:-1, ApplicationImpl (io.quarkus.runner)
ensureClassInitialized0:-1, Unsafe (jdk.internal.misc)
ensureClassInitialized:1160, Unsafe (jdk.internal.misc)
ensureClassInitialized:177, ClassInitializationSupport (com.oracle.svm.hosted.classinitialization)
computeInitKindAndMaybeInitializeClass:191, AllowAllHostedUsagesClassInitializationSupport (com.oracle.svm.hosted.classinitialization)
computeInitKindAndMaybeInitializeClass:129, AllowAllHostedUsagesClassInitializationSupport (com.oracle.svm.hosted.classinitialization)
maybeInitializeAtBuildTime:161, ClassInitializationSupport (com.oracle.svm.hosted.classinitialization)
maybeInitializeAtBuildTime:150, ClassInitializationSupport (com.oracle.svm.hosted.classinitialization)
onTypeReachable:310, SVMHost (com.oracle.svm.hosted)
onTypeReachable:699, AnalysisUniverse (com.oracle.graal.pointsto.meta)
lambda$new$0:310, AnalysisType (com.oracle.graal.pointsto.meta)
run:-1, AnalysisType$$Lambda/0x00000008028590c0 (com.oracle.graal.pointsto.meta)
call:572, Executors$RunnableAdapter (java.util.concurrent)
run:317, FutureTask (java.util.concurrent)
ensureDone:63, AnalysisFuture (com.oracle.graal.pointsto.util)
ensureOnTypeReachableTaskDone:696, AnalysisType (com.oracle.graal.pointsto.meta)
onReachable:590, AnalysisType (com.oracle.graal.pointsto.meta)
run:-1, AnalysisType$$Lambda/0x0000000802859950 (com.oracle.graal.pointsto.meta)
atomicSetAndRun:49, AtomicUtils (com.oracle.graal.pointsto.util)
lambda$registerAsReachable$8:562, AnalysisType (com.oracle.graal.pointsto.meta)
accept:-1, AnalysisType$$Lambda/0x0000000802859720 (com.oracle.graal.pointsto.meta)
forAllSuperTypes:676, AnalysisType (com.oracle.graal.pointsto.meta)
forAllSuperTypes:659, AnalysisType (com.oracle.graal.pointsto.meta)
forAllSuperTypes:655, AnalysisType (com.oracle.graal.pointsto.meta)
registerAsReachable:562, AnalysisType (com.oracle.graal.pointsto.meta)
registerClass:204, ReflectionDataBuilder (com.oracle.svm.hosted.reflect)
lambda$register$2:173, ReflectionDataBuilder (com.oracle.svm.hosted.reflect)
run:-1, ReflectionDataBuilder$$Lambda/0x0000000802914220 (com.oracle.svm.hosted.reflect)
executeCommand:187, CompletionExecutor (com.oracle.graal.pointsto.util)
lambda$executeService$0:171, CompletionExecutor (com.oracle.graal.pointsto.util)
run:-1, CompletionExecutor$$Lambda/0x00000008029a5618 (com.oracle.graal.pointsto.util)
exec:1423, ForkJoinTask$RunnableExecuteAction (java.util.concurrent)
doExec:387, ForkJoinTask (java.util.concurrent)
externalHelpQuiesce:2260, ForkJoinPool (java.util.concurrent)
helpQuiescePool:2297, ForkJoinPool (java.util.concurrent)
awaitQuiescence:3565, ForkJoinPool (java.util.concurrent)
complete:237, CompletionExecutor (com.oracle.graal.pointsto.util)
doTypeflow:529, PointsToAnalysis (com.oracle.graal.pointsto)
finish:517, PointsToAnalysis (com.oracle.graal.pointsto)
runAnalysis:162, AbstractAnalysisEngine (com.oracle.graal.pointsto)
runPointsToAnalysis:783, NativeImageGenerator (com.oracle.svm.hosted)
doRun:592, NativeImageGenerator (com.oracle.svm.hosted)
run:550, NativeImageGenerator (com.oracle.svm.hosted)
buildImage:539, NativeImageGeneratorRunner (com.oracle.svm.hosted)
build:721, NativeImageGeneratorRunner (com.oracle.svm.hosted)
start:143, NativeImageGeneratorRunner (com.oracle.svm.hosted)
main:98, NativeImageGeneratorRunner (com.oracle.svm.hosted)
```
</details>

<details>
<summary>`io.quarkus.arc.Arc.container` is first invoked through:</summary>

```
container:56, Arc (io.quarkus.arc)
<init>:39, ArcCDIProvider$ArcCDI (io.quarkus.arc.impl) // <----------------------
<init>:23, ArcCDIProvider (io.quarkus.arc.impl)
init:273, ArcContainerImpl (io.quarkus.arc.impl)
initialize:40, Arc (io.quarkus.arc)
initContainer:49, ArcRecorder (io.quarkus.arc.runtime)
deploy_0:-1, ArcProcessor$initializeContainer1770303700 (io.quarkus.deployment.steps)
deploy:-1, ArcProcessor$initializeContainer1770303700 (io.quarkus.deployment.steps)
<clinit>:-1, ApplicationImpl (io.quarkus.runner)
ensureClassInitialized0:-1, Unsafe (jdk.internal.misc)
ensureClassInitialized:1160, Unsafe (jdk.internal.misc)
ensureClassInitialized:177, ClassInitializationSupport (com.oracle.svm.hosted.classinitialization)
computeInitKindAndMaybeInitializeClass:191, AllowAllHostedUsagesClassInitializationSupport (com.oracle.svm.hosted.classinitialization)
computeInitKindAndMaybeInitializeClass:129, AllowAllHostedUsagesClassInitializationSupport (com.oracle.svm.hosted.classinitialization)
maybeInitializeAtBuildTime:161, ClassInitializationSupport (com.oracle.svm.hosted.classinitialization)
maybeInitializeAtBuildTime:150, ClassInitializationSupport (com.oracle.svm.hosted.classinitialization)
onTypeReachable:310, SVMHost (com.oracle.svm.hosted)
onTypeReachable:699, AnalysisUniverse (com.oracle.graal.pointsto.meta)
lambda$new$0:310, AnalysisType (com.oracle.graal.pointsto.meta)
run:-1, AnalysisType$$Lambda/0x000000080281b780 (com.oracle.graal.pointsto.meta)
call:572, Executors$RunnableAdapter (java.util.concurrent)
run:317, FutureTask (java.util.concurrent)
ensureDone:63, AnalysisFuture (com.oracle.graal.pointsto.util)
ensureOnTypeReachableTaskDone:696, AnalysisType (com.oracle.graal.pointsto.meta)
onReachable:590, AnalysisType (com.oracle.graal.pointsto.meta)
run:-1, AnalysisType$$Lambda/0x000000080281bde0 (com.oracle.graal.pointsto.meta)
atomicSetAndRun:49, AtomicUtils (com.oracle.graal.pointsto.util)
lambda$registerAsReachable$8:562, AnalysisType (com.oracle.graal.pointsto.meta)
accept:-1, AnalysisType$$Lambda/0x00000008028f6000 (com.oracle.graal.pointsto.meta)
forAllSuperTypes:676, AnalysisType (com.oracle.graal.pointsto.meta)
forAllSuperTypes:659, AnalysisType (com.oracle.graal.pointsto.meta)
forAllSuperTypes:655, AnalysisType (com.oracle.graal.pointsto.meta)
registerAsReachable:562, AnalysisType (com.oracle.graal.pointsto.meta)
registerClass:204, ReflectionDataBuilder (com.oracle.svm.hosted.reflect)
lambda$register$2:173, ReflectionDataBuilder (com.oracle.svm.hosted.reflect)
run:-1, ReflectionDataBuilder$$Lambda/0x00000008029126f0 (com.oracle.svm.hosted.reflect)
executeCommand:187, CompletionExecutor (com.oracle.graal.pointsto.util)
lambda$executeService$0:171, CompletionExecutor (com.oracle.graal.pointsto.util)
run:-1, CompletionExecutor$$Lambda/0x00000008029a5618 (com.oracle.graal.pointsto.util)
exec:1423, ForkJoinTask$RunnableExecuteAction (java.util.concurrent)
doExec:387, ForkJoinTask (java.util.concurrent)
externalHelpQuiesce:2260, ForkJoinPool (java.util.concurrent)
helpQuiescePool:2297, ForkJoinPool (java.util.concurrent)
awaitQuiescence:3565, ForkJoinPool (java.util.concurrent)
complete:237, CompletionExecutor (com.oracle.graal.pointsto.util)
doTypeflow:529, PointsToAnalysis (com.oracle.graal.pointsto)
finish:517, PointsToAnalysis (com.oracle.graal.pointsto)
runAnalysis:162, AbstractAnalysisEngine (com.oracle.graal.pointsto)
runPointsToAnalysis:783, NativeImageGenerator (com.oracle.svm.hosted)
doRun:592, NativeImageGenerator (com.oracle.svm.hosted)
run:550, NativeImageGenerator (com.oracle.svm.hosted)
buildImage:539, NativeImageGeneratorRunner (com.oracle.svm.hosted)
build:721, NativeImageGeneratorRunner (com.oracle.svm.hosted)
start:143, NativeImageGeneratorRunner (com.oracle.svm.hosted)
main:98, NativeImageGeneratorRunner (com.oracle.svm.hosted)
```
</details>

Looking at the top of the stack traces we see that `io.quarkus.arc.Arc.container` is always invoked (sequentially) before the instantiation of `io.quarkus.arc.impl.BeanManagerImpl#INSTANCE` (line 39 vs line 40 in https://github.com/quarkusio/quarkus/blob/6543030ea3073b732a91167c484ed3849dd0d90b/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/ArcCDIProvider.java#L38-L42), as a result the parsing of `io.quarkus.arc.Arc.container` is expected to have been completed before the instantiation of `io.quarkus.arc.impl.BeanManagerImpl#INSTANCE`.

Note however that there appears to be nothing preventing `io.quarkus.arc.impl.BeanManagerImpl` to gets build time initialized independently, although according to my tests it does not happen.

`io.quarkus.arc.impl.BeanManagerImpl#INSTANCE` can be accessed from two places:

https://github.com/quarkusio/quarkus/blob/6543030ea3073b732a91167c484ed3849dd0d90b/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/BeanManagerProvider.java#L17

and 

https://github.com/quarkusio/quarkus/blob/6543030ea3073b732a91167c484ed3849dd0d90b/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/ArcContainerImpl.java#L404

If any of these code-paths is reached before https://github.com/quarkusio/quarkus/blob/6543030ea3073b732a91167c484ed3849dd0d90b/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/ArcCDIProvider.java#L39

then the reported issue could manifest.

I tried adding an artificial delay of 50s before the invocation of  `io.quarkus.arc.Arc.container` and I still couldn't get `io.quarkus.arc.impl.BeanManagerImpl#INSTANCE` to instantiate before it, so there seems to be some implicit synchronization in place preventing this.

**Comment by rsvoboda** [comment]
also cc @jerboaa

**Comment by mkouba** [comment]
I still don't understand 

> Looks like we could potentially solve this by resetting `io.quarkus.arc.impl.ComputingCache.map`.

We can't just reset the CDI singleton context, if that's what you mean...

**Comment by jerboaa** [comment]
> > Looks like we could potentially solve this by resetting `io.quarkus.arc.impl.ComputingCache.map`.
> 
> We can't just reset the CDI singleton context, if that's what you mean...

I meant `@RecomputeFieldValue(kind = RecomputeFieldValue.Kind.Reset)` via Graal substitutions. Similar to this one:
https://github.com/quarkusio/quarkus/blob/9ce6a231fe4bcc93cade531203a80ee8d0a16080/extensions/avro/runtime/src/main/java/io/quarkus/avro/runtime/graal/AvroSubstitutions.java#L32-L34

**Comment by jerboaa** [comment]
But thinking some more about it, you'd probably want to runtime-init that class instead?

**Comment by mkouba** [comment]
> But thinking some more about it, you'd probably want to runtime-init that class instead?

Neither `@RecomputeFieldValue(kind = RecomputeFieldValue.Kind.Reset)` nor runtime-init that class. The CDI singleton context is intentionally available during STATIC_INIT so that `@Singleton` CDI  beans initialized during STATIC_INIT keep the state for the runtime...

**Comment by galderz** [comment]
FYI @ppalaga has just encountered this issue, see https://github.com/quarkiverse/quarkus-cxf/issues/1856
